File: conformance/ts/models/cbor.mjs
// Independent deterministic-CBOR codec per doc/KERNEL.md §3 — written
// against the contract prose, not ported from zerodb-core/src/cbor.rs.
// Values: {t:'uint'|'bytes'|'text'|'array'|'map'|'bool'|'null', ...} tagged
// form (the same shape vectors use).

const MAX_DEPTH = 16;

function hexToBytes(hex) {
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i += 1) out[i] = parseInt(hex.substr(i * 2, 2), 16);
  return out;
}

export function bytesToHex(bytes) {
  return [...bytes].map((b) => b.toString(16).padStart(2, '0')).join('');
}

function header(major, arg) {
  const mt = major << 5;
  if (arg < 24) return [mt | Number(arg)];
  if (arg < 0x100) return [mt | 24, Number(arg)];
  if (arg < 0x10000) return [mt | 25, Number(arg) >> 8, Number(arg) & 0xff];
  if (arg < 0x100000000) {
    const n = Number(arg);
    return [mt | 26, (n >>> 24) & 0xff, (n >>> 16) & 0xff, (n >>> 8) & 0xff, n & 0xff];
  }
  const big = BigInt(arg);
  const bytes = [mt | 27];
  for (let shift = 56n; shift >= 0n; shift -= 8n) bytes.push(Number((big >> shift) & 0xffn));
  return bytes;
}

function encodeTextKey(key) {
  const utf8 = new TextEncoder().encode(key);
  return [...header(3, utf8.length), ...utf8];
}

function encodeInto(value, out, depth) {
  if (depth > MAX_DEPTH) throw new Error('Depth');
  switch (value.t) {
    case 'uint':
      out.push(...header(0, value.v));
      break;
    case 'bytes': {
      const bytes = hexToBytes(value.hex);
      out.push(...header(2, bytes.length), ...bytes);
      break;
    }
    case 'text': {
      const utf8 = new TextEncoder().encode(value.v);
      out.push(...header(3, utf8.length), ...utf8);
      break;
    }
    case 'array':
      out.push(...header(4, value.v.length));
      for (const item of value.v) encodeInto(item, out, depth + 1);
      break;
    case 'map': {
      const entries = Object.entries(value.v).map(([k, v]) => [encodeTextKey(k), v]);
      entries.sort((a, b) => compareBytes(a[0], b[0]));
      for (let i = 1; i < entries.length; i += 1) {
        if (compareBytes(entries[i - 1][0], entries[i][0]) === 0) throw new Error('DuplicateKey');
      }
      out.push(...header(5, entries.length));
      for (const [keyBytes, v] of entries) {
        out.push(...keyBytes);
        encodeInto(v, out, depth + 1);
      }
      break;
    }
    case 'bool':
      out.push(value.v ? 0xf5 : 0xf4);
      break;
    case 'null':
      out.push(0xf6);
      break;
    default:
      throw new Error(`unknown tagged type ${value.t}`);
  }
}

function compareBytes(a, b) {
  const len = Math.min(a.length, b.length);
  for (let i = 0; i < len; i += 1) {
    if (a[i] !== b[i]) return a[i] - b[i];
  }
  return a.length - b.length;
}

export function encode(value) {
  const out = [];
  encodeInto(value, out, 0);
  return new Uint8Array(out);
}

// ------------------------------------------------------------- decoder

class Cursor {
  constructor(bytes) {
    this.bytes = bytes;
    this.pos = 0;
  }

  byte() {
    if (this.pos >= this.bytes.length) throw new Error('Truncated');
    return this.bytes[this.pos++];
  }

  take(n) {
    if (this.pos + n > this.bytes.length) throw new Error('Truncated');
    const slice = this.bytes.subarray(this.pos, this.pos + n);
    this.pos += n;
    return slice;
  }

  arg(additional) {
    if (additional < 24) return additional;
    if (additional === 24) {
      const v = this.byte();
      if (v < 24) throw new Error('NonCanonical');
      return v;
    }
    if (additional === 25) {
      const b = this.take(2);
      const v = (b[0] << 8) | b[1];
      if (v < 0x100) throw new Error('NonCanonical');
      return v;
    }
    if (additional === 26) {
      const b = this.take(4);
      const v = b[0] * 0x1000000 + (b[1] << 16) + (b[2] << 8) + b[3];
      if (v < 0x10000) throw new Error('NonCanonical');
      return v;
    }
    if (additional === 27) {
      const b = this.take(8);
      let v = 0n;
      for (const byte of b) v = (v << 8n) | BigInt(byte);
      if (v < 0x100000000n) throw new Error('NonCanonical');
      return v <= BigInt(Number.MAX_SAFE_INTEGER) ? Number(v) : v;
    }
    throw new Error('NonCanonical'); // 28-30 reserved, 31 indefinite
  }

  item(depth) {
    if (depth > MAX_DEPTH) throw new Error('Depth');
    const initial = this.byte();
    const major = initial >> 5;
    const additional = initial & 0x1f;
    switch (major) {
      case 0:
        return { t: 'uint', v: this.arg(additional) };
      case 2: {
        const len = Number(this.arg(additional));
        return { t: 'bytes', hex: bytesToHex(this.take(len)) };
      }
      case 3: {
        const len = Number(this.arg(additional));
        return { t: 'text', v: new TextDecoder('utf-8', { fatal: true }).decode(this.take(len)) };
      }
      case 4: {
        const len = Number(this.arg(additional));
        const items = [];
        for (let i = 0; i < len; i += 1) items.push(this.item(depth + 1));
        return { t: 'array', v: items };
      }
      case 5: {
        const len = Number(this.arg(additional));
        const obj = {};
        let prevKey = null;
        for (let i = 0; i < len; i += 1) {
          const keyStart = this.pos;
          const key = this.item(depth + 1);
          if (key.t !== 'text') throw new Error('Unsupported');
          const keyBytes = this.bytes.subarray(keyStart, this.pos);
          if (prevKey !== null) {
            const cmp = compareBytes(prevKey, keyBytes);
            if (cmp === 0) throw new Error('DuplicateKey');
            if (cmp > 0) throw new Error('KeyOrder');
          }
          prevKey = keyBytes;
          obj[key.v] = this.item(depth + 1);
        }
        return { t: 'map', v: obj };
      }
      case 7:
        if (initial === 0xf4) return { t: 'bool', v: false };
        if (initial === 0xf5) return { t: 'bool', v: true };
        if (initial === 0xf6) return { t: 'null' };
        throw new Error('Unsupported');
      default:
        throw new Error('Unsupported'); // major 1 (nint) unused; 6 = tags
    }
  }
}

export function decode(bytes) {
  const cursor = new Cursor(bytes);
  let value;
  try {
    value = cursor.item(0);
  } catch (e) {
    if (e instanceof RangeError || e.name === 'TypeError') throw new Error('Utf8');
    throw e;
  }
  if (cursor.pos !== bytes.length) throw new Error('Trailing');
  return value;
}

export { hexToBytes };

File: zerodb-core/src/cbor.rs
//! Deterministic CBOR subset per doc/KERNEL.md §3.
//!
//! RFC 8949 Core Deterministic Encoding, further restricted: definite
//! lengths only; text-string map keys, unique, sorted bytewise by encoded
//! form; shortest-form integers; no floats; no tags; simple values limited
//! to false/true/null. The decoder REJECTS non-canonical input — decode →
//! re-encode is byte-identical by construction (INVARIANTS I-6).

use thiserror::Error;

/// Maximum nesting depth (registry `limits.max_cbor_depth`).
pub const MAX_DEPTH: usize = 16;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Cbor {
    Uint(u64),
    Bytes(Vec<u8>),
    Text(String),
    Array(Vec<Cbor>),
    /// Entries in encoded-key order; `encode` sorts, `decode` verifies.
    Map(Vec<(String, Cbor)>),
    Bool(bool),
    Null,
}

#[derive(Debug, Error, PartialEq, Eq)]
pub enum CborError {
    #[error("duplicate map key")]
    DuplicateKey,
    #[error("map keys not in canonical order")]
    KeyOrder,
    #[error("nesting depth exceeds {MAX_DEPTH}")]
    Depth,
    #[error("non-canonical encoding (non-shortest form or indefinite length)")]
    NonCanonical,
    #[error("unsupported CBOR item (floats, tags, and other simples are excluded)")]
    Unsupported,
    #[error("truncated input")]
    Truncated,
    #[error("trailing bytes after top-level item")]
    Trailing,
    #[error("invalid UTF-8 in text string")]
    Utf8,
}

// ---------------------------------------------------------------- encode

fn write_header(out: &mut Vec<u8>, major: u8, arg: u64) {
    let mt = major << 5;
    match arg {
        0..=23 => out.push(mt | arg as u8),
        24..=0xff => {
            out.push(mt | 24);
            out.push(arg as u8);
        }
        0x100..=0xffff => {
            out.push(mt | 25);
            out.extend_from_slice(&(arg as u16).to_be_bytes());
        }
        0x1_0000..=0xffff_ffff => {
            out.push(mt | 26);
            out.extend_from_slice(&(arg as u32).to_be_bytes());
        }
        _ => {
            out.push(mt | 27);
            out.extend_from_slice(&arg.to_be_bytes());
        }
    }
}

fn encode_text_key(key: &str) -> Vec<u8> {
    let mut out = Vec::with_capacity(key.len() + 2);
    write_header(&mut out, 3, key.len() as u64);
    out.extend_from_slice(key.as_bytes());
    out
}

fn encode_into(value: &Cbor, out: &mut Vec<u8>, depth: usize) -> Result<(), CborError> {
    if depth > MAX_DEPTH {
        return Err(CborError::Depth);
    }
    match value {
        Cbor::Uint(n) => write_header(out, 0, *n),
        Cbor::Bytes(b) => {
            write_header(out, 2, b.len() as u64);
            out.extend_from_slice(b);
        }
        Cbor::Text(s) => {
            write_header(out, 3, s.len() as u64);
            out.extend_from_slice(s.as_bytes());
        }
        Cbor::Array(items) => {
            write_header(out, 4, items.len() as u64);
            for item in items {
                encode_into(item, out, depth + 1)?;
            }
        }
        Cbor::Map(entries) => {
            let mut encoded: Vec<(Vec<u8>, &Cbor)> = entries
                .iter()
                .map(|(k, v)| (encode_text_key(k), v))
                .collect();
            encoded.sort_by(|a, b| a.0.cmp(&b.0));
            for pair in encoded.windows(2) {
                if pair[0].0 == pair[1].0 {
                    return Err(CborError::DuplicateKey);
                }
            }
            write_header(out, 5, encoded.len() as u64);
            for (key_bytes, v) in encoded {
                out.extend_from_slice(&key_bytes);
                encode_into(v, out, depth + 1)?;
            }
        }
        Cbor::Bool(false) => out.push(0xf4),
        Cbor::Bool(true) => out.push(0xf5),
        Cbor::Null => out.push(0xf6),
    }
    Ok(())
}

/// Canonical encoding of `value`.
pub fn encode(value: &Cbor) -> Result<Vec<u8>, CborError> {
    let mut out = Vec::new();
    encode_into(value, &mut out, 0)?;
    Ok(out)
}

// ---------------------------------------------------------------- decode

struct Cursor<'a> {
    input: &'a [u8],
    pos: usize,
}

impl<'a> Cursor<'a> {
    fn byte(&mut self) -> Result<u8, CborError> {
        let b = *self.input.get(self.pos).ok_or(CborError::Truncated)?;
        self.pos += 1;
        Ok(b)
    }

    fn take(&mut self, n: usize) -> Result<&'a [u8], CborError> {
        let end = self.pos.checked_add(n).ok_or(CborError::Truncated)?;
        if end > self.input.len() {
            return Err(CborError::Truncated);
        }
        let slice = &self.input[self.pos..end];
        self.pos = end;
        Ok(slice)
    }

    /// Read a length/value argument, enforcing shortest form.
    fn arg(&mut self, additional: u8) -> Result<u64, CborError> {
        match additional {
            0..=23 => Ok(additional as u64),
            24 => {
                let v = self.byte()? as u64;
                if v < 24 {
                    return Err(CborError::NonCanonical);
                }
                Ok(v)
            }
            25 => {
                let v = u16::from_be_bytes(self.take(2)?.try_into().unwrap()) as u64;
                if v < 0x100 {
                    return Err(CborError::NonCanonical);
                }
                Ok(v)
            }
            26 => {
                let v = u32::from_be_bytes(self.take(4)?.try_into().unwrap()) as u64;
                if v < 0x1_0000 {
                    return Err(CborError::NonCanonical);
                }
                Ok(v)
            }
            27 => {
                let v = u64::from_be_bytes(self.take(8)?.try_into().unwrap());
                if v < 0x1_0000_0000 {
                    return Err(CborError::NonCanonical);
                }
                Ok(v)
            }
            // 28-30 reserved, 31 = indefinite length: both non-canonical here.
            _ => Err(CborError::NonCanonical),
        }
    }

    fn item(&mut self, depth: usize) -> Result<Cbor, CborError> {
        if depth > MAX_DEPTH {
            return Err(CborError::Depth);
        }
        let initial = self.byte()?;
        let major = initial >> 5;
        let additional = initial & 0x1f;
        match major {
            0 => Ok(Cbor::Uint(self.arg(additional)?)),
            2 => {
                let len = self.arg(additional)? as usize;
                Ok(Cbor::Bytes(self.take(len)?.to_vec()))
            }
            3 => {
                let len = self.arg(additional)? as usize;
                let s = std::str::from_utf8(self.take(len)?).map_err(|_| CborError::Utf8)?;
                Ok(Cbor::Text(s.to_owned()))
            }
            4 => {
                let len = self.arg(additional)? as usize;
                let mut items = Vec::with_capacity(len.min(1024));
                for _ in 0..len {
                    items.push(self.item(depth + 1)?);
                }
                Ok(Cbor::Array(items))
            }
            5 => {
                let len = self.arg(additional)? as usize;
                let mut entries: Vec<(String, Cbor)> = Vec::with_capacity(len.min(1024));
                let mut prev_key: Option<Vec<u8>> = None;
                for _ in 0..len {
                    let key_start = self.pos;
                    let key = match self.item(depth + 1)? {
                        Cbor::Text(k) => k,
                        _ => return Err(CborError::Unsupported),
                    };
                    let key_bytes = self.input[key_start..self.pos].to_vec();
                    if let Some(prev) = &prev_key {
                        if key_bytes == *prev {
                            return Err(CborError::DuplicateKey);
                        }
                        if key_bytes < *prev {
                            return Err(CborError::KeyOrder);
                        }
                    }
                    prev_key = Some(key_bytes);
                    entries.push((key, self.item(depth + 1)?));
                }
                Ok(Cbor::Map(entries))
            }
            7 => match initial {
                0xf4 => Ok(Cbor::Bool(false)),
                0xf5 => Ok(Cbor::Bool(true)),
                0xf6 => Ok(Cbor::Null),
                _ => Err(CborError::Unsupported),
            },
            // major 1 (negative ints) unused by the kernel; 6 = tags.
            _ => Err(CborError::Unsupported),
        }
    }
}

/// Decode one canonical top-level item; rejects trailing bytes.
pub fn decode(input: &[u8]) -> Result<Cbor, CborError> {
    let mut cursor = Cursor { input, pos: 0 };
    let value = cursor.item(0)?;
    if cursor.pos != input.len() {
        return Err(CborError::Trailing);
    }
    Ok(value)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trip_is_byte_identical() {
        let value = Cbor::Map(vec![
            ("zz".into(), Cbor::Uint(300)),
            (
                "a".into(),
                Cbor::Array(vec![Cbor::Bytes(vec![1, 2]), Cbor::Null]),
            ),
        ]);
        let bytes = encode(&value).unwrap();
        let decoded = decode(&bytes).unwrap();
        assert_eq!(encode(&decoded).unwrap(), bytes);
    }

    #[test]
    fn rejects_duplicate_keys() {
        // {"a": 1, "a": 2}
        assert_eq!(
            decode(&[0xa2, 0x61, 0x61, 0x01, 0x61, 0x61, 0x02]),
            Err(CborError::DuplicateKey)
        );
    }

    #[test]
    fn rejects_non_shortest_uint() {
        // 23 encoded with a one-byte argument
        assert_eq!(decode(&[0x18, 0x17]), Err(CborError::NonCanonical));
    }

    #[test]
    fn rejects_out_of_order_keys() {
        // {"b": 1, "a": 2}
        assert_eq!(
            decode(&[0xa2, 0x61, 0x62, 0x01, 0x61, 0x61, 0x02]),
            Err(CborError::KeyOrder)
        );
    }

    #[test]
    fn rejects_floats_and_tags() {
        assert_eq!(decode(&[0xf9, 0x00, 0x00]), Err(CborError::Unsupported)); // f16
        assert_eq!(decode(&[0xc0, 0x00]), Err(CborError::Unsupported)); // tag 0
    }
}
